### PR TITLE
Add deprecation warnings for EOL Rails and Ruby and the associated work arounds

### PR DESCRIPTION
**Pull request by reeganviljoen**
### What are you trying to accomplish?

Add deprecation warnings for EOL Rails and Ruby and the associated work arounds

### CODE REVIEW

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,8 @@ nav_order: 5
 
 ## main
 
+* Add deprecation warnings for EOL ruby and Rails versions and patches assocaited with them
```
Comment (camertron) [inline docs/CHANGELOG.md:13]:
```suggestion
* Add deprecation warnings for EOL ruby and Rails versions and patches associated with them
```

**Review by camertron — APPROVED** [review]

**Comment by joelhawksley** [comment]
@reeganviljoen mind cleaning up the code coverage on this PR, either by adding `nocov` or tests?

File: lib/view_component/engine.rb [hunk 2]
```diff
@@ -79,6 +79,7 @@ class Engine < Rails::Engine # :nodoc:
 
     initializer "view_component.monkey_patch_render" do |app|
       next if Rails.version.to_f >= 6.1 || !app.config.view_component.render_monkey_patch_enabled
+      ViewComponent::Deprecation.deprecation_warning("Monkey patching `render`", "ViewComponent 4.0 will drop monkey patching `render`")
```
Comment (joelhawksley) [inline lib/view_component/engine.rb:82]:
```suggestion
      ViewComponent::Deprecation.deprecation_warning("Monkey patching `render`", "ViewComponent 4.0 will remove the `render` monkey patch")
```

File: lib/view_component/engine.rb [hunk 3]
```diff
@@ -97,6 +98,7 @@ class Engine < Rails::Engine # :nodoc:
 
     initializer "view_component.include_render_component" do |_app|
       next if Rails.version.to_f >= 6.1
+      ViewComponent::Deprecation.deprecation_warning("using `render_component`", "ViewComponent 4.0 will drop using `render_component`")
```
Comment (joelhawksley) [inline lib/view_component/engine.rb:101]:
```suggestion
      ViewComponent::Deprecation.deprecation_warning("using `render_component`", "ViewComponent 4.0 will remove `render_component`")
```

File: lib/view_component/engine.rb [hunk 4]
```diff
@@ -160,6 +162,14 @@ def serve_static_preview_assets?(app_config)
         end
       end
 
+      if RUBY_VERSION < "3.0.0"
+        ViewComponent::Deprecation.deprecation_warning("Support for Ruby versions < 3.0.0", "ViewComponent 4.0 will drop support for Ruby versions < 3.0.0 ")
```
Comment (joelhawksley) [inline lib/view_component/engine.rb:166]:
```suggestion
        ViewComponent::Deprecation.deprecation_warning("Support for Ruby versions < 3.0.0", "ViewComponent 4.0 will remove support for Ruby versions < 3.0.0 ")
```

File: lib/view_component/engine.rb [hunk 5]
```diff
@@ -160,6 +162,14 @@ def serve_static_preview_assets?(app_config)
...
+      end
+
+      if Rails.version.to_f < 6.1
+        ViewComponent::Deprecation.deprecation_warning("Support for Rails versions < 6.1", "ViewComponent 4.0 will drop support for Rails versions < 6.1 ")
```
Comment (joelhawksley) [inline lib/view_component/engine.rb:170]:
```suggestion
        ViewComponent::Deprecation.deprecation_warning("Support for Rails versions < 6.1", "ViewComponent 4.0 will remove support for Rails versions < 6.1 ")
```

**Comment by reeganviljoen** [comment]
@joelhawksley sorry for the late reply, I have some time tommorow, I will lol at it then

**Comment by reeganviljoen** [comment]
@joelhawksley the deprecation warnings are silenced in tests so I opted to use `#nocov`

**Pull request merged**